volumina/utility/preferences.py

```diff
@@ -0,0 +1,155 @@
+
+    If there is no such group, or there is no such setting in the
+    existing group, the default value is returned.
+
+    See Also:
+        :func:`getmany`.
+    """
+    return getmany((group, setting, default))[0]
+
+
+def set(group: str, setting: str, value: Any) -> None:
+    """Set a preference setting in the specified group.
+
+    If the group or the setting in a group does not exist, it will be
+    created.
+
+    See Also:
+        :func:`setmany`.
+    """
+    setmany((group, setting, value))
+
+
+def getmany(*args: Tuple[str, str, Any]) -> Tuple[Any, ...]:
+    """Return multiple preference settings in a single transaction.
+
+    See Also:
+        :func:`get`.
+    """
+    with _lock:
+        try:
+            return _getmany(*args)
+        except Exception:
+            logger.exception("Failed to load preferences from %r", str(_path))
+            _path.unlink()
+            return tuple(default for _group, _setting, default in args)
+
+
+def _getmany(*args: Tuple[str, str, Any]) -> Tuple[Any, ...]:
+    data = _load()
+    result = []
+    for group, setting, default in args:
+        try:
+            result.append(data[group][setting])
+        except KeyError:
+            result.append(default)
+    return tuple(result)
+
+
+def setmany(*args: Tuple[str, str, Any]) -> None:
+    """Set multiple preference settings in a single transaction.
+
+    See Also:
+        :func:`set`.
+    """
+    with _lock:
+        try:
+            _setmany(*args)
+        except Exception:
+            logger.exception("Failed to save preferences to %r", str(_path))
+            _path.unlink()
+
+
+def _setmany(*args: Tuple[str, str, Any]) -> None:
+    data = _load()
+    for group, setting, value in args:
+        data.setdefault(group, {})[setting] = value
+    with open(_path, "wb") as f:
+        # Use the lowest available protocol for backwards compatibility.
+        pickle.dump(data, f, protocol=0)
+
+
+def _load() -> Any:
+    """Read serialized preferences from file.
+
+    The returned value should have the type ``Dict[str, Dict[str, Any]]``,
+    but ``Any`` is used instead because deserialization can return any
+    valid Python object. No type validation is performed because
+    callers might need only a single value from the composite structure.
+    """
+    try:
+        with open(_path, "rb") as f:
+            return pickle.load(f)
+    except FileNotFoundError:
+        return {}
+
+
+def get_location() -> pathlib.Path:
+    """Return the current preferences file location."""
+    with _lock:
+        return _path
+
+
+def change_location(path: os.PathLike) -> pathlib.Path:
+    """Change the preferences file location.
+
+    All subsequent preferences operations will use the specified file.
+    This function simply changes the source for reads and writes, it
+    does not copy the contents of the previous file into a new one.
+
+    Returns:
+        Old preferences file location.
+    """
+    with _lock:
```

Comment:
I think this `_lock` is not doing much locking at all... It doesn't protect a client from multiple other clients changing the `_path`. Say we have two clients like so:

```
preferences.change_location("first_location") # client 1
preferences.set("some_group", "some_property", 123) # client 1
```

```
preferences.change_location("second_location") #client 2
preferences.set("some_group", "some_property", 456) #client 2
```

then this is a valid execution order:

```
preferences.change_location("first_location") # client 1
preferences.change_location("second_location") #client 2
preferences.set("some_group", "some_property", 123) # client 1
preferences.set("some_group", "some_property", 456) #client 2
```
and even with the locks, everything would be written into `"second_location"`. I see that this has somewhat of an "exchange" semantics, but for that to work the return (old) value would have to have a value that indicates whether you acquired a lock.

Anyway, it might be that we don't even care about having this thread-safe anyway, so we might just do away with the locks altogether.

Reply:
You are right, but I concluded that it would be safer to serialize all accesses to preferences anyway because contention shouldn't be a problem: if you read preferences in a tight loop, then you are doing something wrong.

> Anyway, it might be that we don't even care about having this thread-safe anyway, so we might just do away with the locks altogether.

This is also correct, as @k-dominik pointed out, we can load data from a file and keep it in memory, while stating that `change_location` is not thread-safe.